### PR TITLE
MinGW: Fix error: cannot convert 'size_t*' to 'int*'

### PR DESCRIPTION
**Pull request by yadij**
Removing another addrinfo as we do.

### CODE REVIEW

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/comm/TcpAcceptor.cc [hunk 1]
```diff
@@ -371,23 +368,21 @@ Comm::TcpAcceptor::acceptInto(Comm::ConnectionPointer &details)
...
     // lookup the local-end details of this new connection
-    Ip::Address::InitAddr(gai);
-    details->local.setEmpty();
-    if (getsockname(sock, gai->ai_addr, &gai->ai_addrlen) != 0) {
+    struct sockaddr_storage localAddr = {};
+    socklen_t localAddrSz = sizeof(localAddr);
+    if (getsockname(sock, reinterpret_cast<struct sockaddr *>(&localAddr), &localAddrSz) != 0) {
         int xerrno = errno;
-        Ip::Address::FreeAddr(gai);
+        details->local.setEmpty();
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/comm/TcpAcceptor.cc:380]: AFAIK this was only done for paranoia reasons. Removing.

File: src/comm/TcpAcceptor.cc [hunk 2]
```diff
@@ -371,23 +368,21 @@ Comm::TcpAcceptor::acceptInto(Comm::ConnectionPointer &details)
...
     details->enterOrphanage();
 
-    details->remote = *gai;
-
     // lookup the local-end details of this new connection
-    Ip::Address::InitAddr(gai);
-    details->local.setEmpty();
-    if (getsockname(sock, gai->ai_addr, &gai->ai_addrlen) != 0) {
+    struct sockaddr_storage localAddr = {};
+    socklen_t localAddrSz = sizeof(localAddr);
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/comm/TcpAcceptor.cc]: Done in 849a25b845

File: src/comm/TcpAcceptor.cc [hunk 3]
```diff
@@ -371,23 +368,21 @@ Comm::TcpAcceptor::acceptInto(Comm::ConnectionPointer &details)
...
-        Ip::Address::FreeAddr(gai);
+        details->local.setEmpty();
         debugs(50, DBG_IMPORTANT, "ERROR: Closing accepted TCP connection after failing to obtain its local IP address" <<
                Debug::Extra << "accepted connection: " << details <<
                Debug::Extra << "getsockname(2) error: " << xstrerr(xerrno));
         return false;
     }
-    details->local = *gai;
-    Ip::Address::FreeAddr(gai);
+    details->local = localAddr;
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/comm/TcpAcceptor.cc]: Done in 849a25b845

File: src/comm/TcpAcceptor.cc [hunk 4]
```diff
@@ -345,16 +345,13 @@ bool
...
     ++statCounter.syscalls.sock.accepts;
-    struct addrinfo *gai = nullptr;
-    Ip::Address::InitAddr(gai);
 
     errcode = 0; // reset local errno copy.
-    const auto rawSock = accept(conn->fd, gai->ai_addr, &gai->ai_addrlen);
+    struct sockaddr_storage remoteAddr = {};
+    socklen_t remoteAddrSz = sizeof(remoteAddr);
+    const auto rawSock = accept(conn->fd, reinterpret_cast<struct sockaddr *>(&remoteAddr), &remoteAddrSz);
     if (rawSock < 0) {
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/comm/TcpAcceptor.cc:353]: Done in 849a25b845

File: src/comm/TcpAcceptor.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/comm/TcpAcceptor.cc]: Done in 849a25b845

**Review by rousskov — APPROVED** [review]
Thanks a lot for this cleanup!

P.S. We should avoid (poor) code duplication that this PR and recent 1800e254 are creating, but I think we should _continue_ to duplicate that code for a little bit longer before there are enough use cases to suggest a proper deduplication solution.

**Review by kinkie — APPROVED** [review]

**Comment by yadij** [comment]
> Thanks a lot for this cleanup!
> 
> P.S. We should avoid (poor) code duplication that this PR and recent [1800e25](https://github.com/squid-cache/squid/commit/1800e254070002fc7ee3f4848e06fdbea00fd2fb) are creating, but I think we should _continue_ to duplicate that code for a little bit longer before there are enough use cases to suggest a proper deduplication solution.

I assume you are referring to the pattern of 2 local variables with a `reinterpret_cast` (if not please correct me). FWIW, I hope long term that we only need to de-duplicate the set/get syscalls and the rest are uncommon enough to be okay without anything special.

**Comment by rousskov** [comment]
> > P.S. We should avoid (poor) code duplication that this PR and recent [1800e25](https://github.com/squid-cache/squid/commit/1800e254070002fc7ee3f4848e06fdbea00fd2fb) are creating, but I think we should _continue_ to duplicate that code for a little bit longer before there are enough use cases to suggest a proper deduplication solution.
 
> I assume you are referring to the pattern of 2 local variables with a `reinterpret_cast` (if not please correct me).

The pattern includes two local variable definition/initialization code, a reinterpret_cast, an assertion, and a conversion to an Ip::Address object (at least).

> FWIW, I hope long term that we only need to de-duplicate the set/get syscalls and the rest are uncommon enough to be okay without anything special.

Yeah, let's see where these initial (and duplicated) changes lead us. I see no reason to discuss the details of that potential future duplication avoidance effort until we have more merged use cases. We are making good progress so far...

**Pull request closed without merging**